runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+              output, element.getTimestamp(), element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        public void outputWithTimestamp(OutputT output, Instant timestamp) {
+          internals.outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs > MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
+          baseContext.sideOutputWithTimestamp(tag, output, element.getTimestamp());
+        }
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure this is correct, but there's no "internals.sideOutputWindowedValue".
+          baseContext.sideOutputWithTimestamp(tag, output, timestamp);
+        }
+      };
+    }
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          throw new UnsupportedOperationException("Splittable DoFn's don't support extra context");
+        }
+
+        @Override
+        public DoFn.InputProvider<InputT> inputProvider() {
+          throw new UnsupportedOperationException("Splittable DoFn's don't support extra context");
+        }
+
+        @Override
+        public DoFn.OutputReceiver<OutputT> outputReceiver() {
+          throw new UnsupportedOperationException("Splittable DoFn's don't support extra context");
+        }
+
+        @Override
+        public TrackerT restrictionTracker() {
+          return tracker;
+        }
+      };
+    }
+  }
+
+  /** Splits the restriction using the given {@link DoFn.SplitRestriction} method. */
+  private static class SplitRestrictionFn<InputT, RestrictionT>
+      extends DoFn<KV<InputT, RestrictionT>, KV<InputT, RestrictionT>> {
+    private final DoFn<InputT, ?> fn;
+    private transient DoFnInvoker<InputT, ?> invoker;
+
+    SplitRestrictionFn(DoFn<InputT, ?> fn) {
+      this.fn = fn;
+    }
+
+    @Setup
+    public void setup() {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      for (RestrictionT part :
```

Comment:
I'd suggest putting the result of `invokeSplitRestriction` into a local variable to improve readability of the loop.